source/extensions/common/wasm/context.cc

```diff
@@ -0,0 +1,212 @@
+#include "common/common/enum_to_int.h"
+#include "common/common/logger.h"
+
+#include "extensions/common/wasm/wasm.h"
+#include "extensions/common/wasm/wasm_state.h"
+#include "extensions/common/wasm/well_known_names.h"
+
+#include "absl/base/casts.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/node_hash_map.h"
+#include "absl/strings/str_cat.h"
+#include "absl/synchronization/mutex.h"
+#include "openssl/bytestring.h"
+#include "openssl/hmac.h"
+#include "openssl/sha.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+bool Buffer::copyTo(WasmBase* wasm, size_t start, size_t length, uint64_t ptr_ptr,
+                    uint64_t size_ptr) const {
+  if (start + length > data_.size()) {
+    return false;
+  }
+  absl::string_view s = data_.substr(start, length);
+  return wasm->copyToPointerSize(s, ptr_ptr, size_ptr);
+}
+
+Wasm* Context::wasm() const { return static_cast<Wasm*>(wasm_); }
+
+void Context::error(absl::string_view message) { throw WasmException(std::string(message)); }
+
+void Context::initializeRoot(WasmBase* wasm, PluginBaseSharedPtr plugin) {
+  ContextBase::initializeRoot(wasm, plugin);
+  root_local_info_ = &std::static_pointer_cast<Plugin>(plugin)->local_info_;
+}
+
+uint64_t Context::getCurrentTimeNanoseconds() {
+  return std::chrono::duration_cast<std::chrono::nanoseconds>(
+             wasm()->time_source_.systemTime().time_since_epoch())
+      .count();
+}
+
+WasmResult Context::log(uint64_t level, absl::string_view message) {
+  switch (static_cast<proxy_wasm::LogLevel>(level)) {
+  case proxy_wasm::LogLevel::trace:
+    ENVOY_LOG(trace, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::debug:
+    ENVOY_LOG(debug, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::info:
+    ENVOY_LOG(info, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::warn:
+    ENVOY_LOG(warn, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::error:
+    ENVOY_LOG(error, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::critical:
+    ENVOY_LOG(critical, "wasm log{}: {}", log_prefix(), message);
+    break;
+  default:
+    return WasmResult::BadArgument;
+  }
+  return WasmResult::Ok;
+}
+
+WasmResult Context::getProperty(absl::string_view path, std::string* result) {
+  if (path.size() == 15 && !memcmp(reinterpret_cast<const void*>(&*path.begin()),
+                                   reinterpret_cast<const void*>("plugin_root_id\0"), 15)) {
+    std::cerr << "returning " << root_id() << "\n";
+    *result = std::string(root_id());
+    return WasmResult::Ok;
+  }
+  return WasmResult::Unimplemented;
+}
+
+WasmResult Context::defineMetric(MetricType type, absl::string_view name, uint32_t* metric_id_ptr) {
+  auto stat_name = wasm()->stat_name_set_->add(name);
+  if (type == MetricType::Counter) {
```

Comment:
nit: can this just use a switch statement?

Reply:
Done.